recipe/2024-06-private-bicep-registries.md

```diff
@@ -0,0 +1,258 @@
+> **Issue Reference:** https://github.com/radius-project/radius/issues/6917
+
+### Goals
+- Add support to publish bicep recipes to a private bicep registry.
+- Enable support to use bicep recipes stored in private bicep registries.
+- Support all OCI compliant private registries.
+
+### Non goals
+
+- Workload identity based authentication(Will be covered as part of different design) 
+- Configure private OCI registry credentials in Radius via CLI and API.
+
+### User scenarios (optional)
+As an operator I am responsible for maintaining Bicep recipes for use with Radius. Bicep recipes used contains sensitive information, proprietary configurations, or data that should not be shared publicly and its intended for internal use within our organization and have granular control over who can access, contribute to, or modify bicep recipes. So I store the recipes in a private bicep registry. And I would like to use these private registries as template-path while registering a bicep recipe.
+
+
+## User Experience (if applicable)
+**Sample Input:**
+Command to publish recipe to private bicep registry.
+```diff
+-rad bicep publish --file ./redis-test.bicep --target br:ghcr.io/myregistry/redis-test:v1
+
++rad bicep publish --file ./redis-test.bicep --target br:ghcr.io/myregistry/redis-test:v1 --username <user> --password <password>
+```
+
+**Sample Output:**
+
+No change in the output
+
+
+## Design
+### Design details
+Currently, OCI-compliant registries are used to store Bicep recipes, with the ORAS client facilitating operations like publish and pull these recipes from the registries. ORAS provides package auth, enabling secure client authentication to remote registries. Private Registry credentials information i.e username and password must be stored in a Application.Core/secretStores resource and the secret resource ID is added to the recipe configuration.
+
+During the recipe deployment bicep driver checks for secrets information for that container registry the recipe deploying is stored on,  and creates an oras auth client to authenticate the private bicep registries.
+
+
+e.g: creating a secret store with keys "username" and "password".
+
+```
+resource secretStore 'Applications.Core/secretStores@2023-10-01-preview' = {
+  name: 'acrsecrets'
+  properties:{
+    type: 'generic'
+    resource: 'registry-secrets/acr'
+    data: {
+      'username': {
+        value: '<username>'
+      }
+      'password': {
+        value: '<registry-password>'
+      }
+    }
+  }
+}
+```
+Update the recipe config in the environment to point to this secret resource id.
+```
+"recipeConfig": {
+  "terraform": {
+    ...
+  },
+  "bicep": {
+    "authentication":{
+      "test.azurecr.io":{
+        "secret": "/planes/radius/local/resourcegroups/default/providers/Applications.Core/secretStores/acrsecrets"
+      },
+    }
+  }
+  "env": {
+    ...
+  }
+},
+```
+
+#### Different container registry providers provide different authenticate methods to login.
+
```

Comment:
I'd like to understand more about how this is going to work, and what you think the right scope is. Some questions:

- What's going to be supported in the first iteration of this feature? All of these registries? All of the authentication methods discussed here?
- Do we have to special case any of these registries? Put another way do we need to have "if is ACR" in our code anywhere?
- What do we expect in the secrets for these registries? `username` and `password`? Something else?

Reply:
- What's going to be supported in the first iteration of this feature? All of these registries? All of the authentication methods discussed here?
        - First iteration of feature will have support for username-password for authentication which enables all the methods discussed here and most of the OCI complaint registries.

-Do we have to special case any of these registries? Put another way do we need to have "if is ACR" in our code anywhere?
         - No,  we don't have to write registry provider specific code with username-password based auth, i foresee we may have azure/aws specific code when we have support for federated auth.
-What do we expect in the secrets for these registries? username and password? Something else?
          - yes its just username, password for now.

Reply:
These are all good answers 👍 - please make sure they make it into the doc with the same level of clarity.